src/history/hs_rec.c

```diff
@@ -1008,27 +929,62 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
 }
 
 /*
- * __hs_delete_key_from_pos --
- *     Delete an entire key's worth of data in the history store assuming that the input cursor is
- *     positioned at the beginning of the key range.
+ * __hs_reinsert_key_from_pos --
+ *     Reinsert an entire key's worth of data in the history store assuming that the input cursor is
+ *     positioned at the beginning of the key range. The reinserted values will have 0 start and
+ *     stop timestamps ensure that they are only visible to transaction id based readers.
  */
 static int
-__hs_delete_key_from_pos(
+__hs_reinsert_key_from_pos(
   WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btree_id, const WT_ITEM *key)
 {
+    WT_CURSOR *insert_cursor;
     WT_CURSOR_BTREE *hs_cbt;
     WT_DECL_RET;
-    WT_ITEM hs_key;
+    WT_HS_TIME_POINT start_time_point, stop_time_point;
+    WT_ITEM hs_key, hs_value;
     WT_UPDATE *upd;
-    wt_timestamp_t hs_start_ts;
-    uint64_t hs_counter;
+    wt_timestamp_t durable_timestamp, hs_start_ts, hs_stop_durable_ts;
+    uint64_t hs_counter, hs_insert_counter, hs_upd_type;
     uint32_t hs_btree_id;
     int cmp;
+    const char *open_cursor_cfg[] = {WT_CONFIG_BASE(session, WT_SESSION_open_cursor), NULL};
 
     hs_cbt = (WT_CURSOR_BTREE *)hs_cursor;
+    WT_CLEAR(hs_key);
+    WT_CLEAR(hs_value);
     upd = NULL;
+    insert_cursor = NULL;
+
+    /*
+     * Determine the starting value of our counter, i.e. highest counter value of the timestamp
+     * range for timestamp 0. We'll be inserting at timestamp 0 and don't want to overwrite a
+     * currently existing counter.
+     *
+     * The cursor will also be positioned at the start of the range that we wish to start inserting.
+     */
+    WT_WITHOUT_DHANDLE(
+      session, ret = __wt_open_cursor(session, WT_HS_URI, NULL, open_cursor_cfg, &insert_cursor));
+
+    F_SET(&((WT_CURSOR_BTREE *)insert_cursor)->iface, WT_CURSTD_IGNORE_TOMBSTONE);
```

Comment:
You shouldn't need the cast here. This flag is for `WT_CURSOR`.
Actually, this code is probably setting the bit on the wrong `flags` variable.

Reply:
I have a feeling that since you're only ever inserting with this cursor, you won't need this flag anyway.

Reply:
Ah actually. It's setting the correct flags variable. But it's casting a `WT_CURSOR *` to a `WT_CURSOR_BTREE *` and then taking its `iface` which is essentially casting it back.

It should just be `F_SET(insert_cursor, WT_CURSTD_IGNORE_TOMBSTONE)`.

Reply:
036ed8b